lib/services/downloads_service.dart

```diff
@@ -0,0 +1,1415 @@
+                  case "image/png":
+                    extension = ".png";
+                  case "image/gif":
+                    extension = ".gif";
+                  case "image/webp":
+                    extension = ".webp";
+                }
+                Future.sync(() async {
+                  assert(
+                      listener.file?.path == await event.task.filePath() ||
+                          (extension != null &&
+                              listener.file?.path.replaceFirst(
+                                      RegExp(r'\.image$'), extension) ==
+                                  await event.task.filePath()),
+                      "${listener.name} ${listener.path} ${listener.fileDownloadLocation} ${listener.file?.path} ${await event.task.filePath()} $extension");
+                });
+                if (extension != null &&
+                    listener.file!.path.endsWith(".image")) {
+                  // Do not wait for file move to complete to prevent slowing isar write
+                  unawaited(File(listener.file!.path)
+                      .rename(listener.file!.path
+                          .replaceFirst(RegExp(r'\.image$'), extension))
+                      .then((_) => null,
+                          onError: (e) => GlobalSnackbar.error(e)));
+                  listener.path = listener.path!
+                      .replaceFirst(RegExp(r'\.image$'), extension);
+                }
+              }
+              if (newState == DownloadItemState.failed) {
+                if (event.exception is TaskFileSystemException ||
+                    (event.exception?.description
+                            .contains(RegExp(r'No space left on device')) ??
+                        false)) {
+                  // Retry items that failed from a full filesystem once the user
+                  // cleans it up and restarts/resyncs
+                  newState = DownloadItemState.enqueued;
+                  if (!_fileSystemFull) {
+                    _fileSystemFull = true;
+                    GlobalSnackbar.message((scaffold) =>
+                        AppLocalizations.of(scaffold)!.filesystemFull);
+                  }
+                } else if (event.exception is TaskConnectionException) {
+                  // Retry items with connection errors
+                  newState = DownloadItemState.enqueued;
+                  incrementConnectionErrors(weight: 2);
+                } else if (event.exception != null) {
+                  _downloadsLogger.warning(
+                      "Exception ${event.exception} when downloading ${listener.name}");
+                } else {
+                  _downloadsLogger.warning(
+                      "Received failed download task ${event.toJson()}");
+                }
+              }
+              // Canceled items are expected to have their status updated by the
+              // cancling code.  Cancelled items not handled and left in downloading
```

Comment:
I'm not sure what you mean about the whitespace.

Reply:
I think you always put a tab (2 spaces) behind the first sentences in comments, not sure why. Here it would be `cancling code.++Cancelled`, where `+` is a space. But since it's consistent across all files and comments, you might as well leave it like this ^^